### PR TITLE
refactor(connect): brand backend descriptors as CanonicalDescriptor

### PR DESCRIPTION
**Pull request by mroz22**
## Summary

Stacked on top of #32328 (`feat_taproot_h_descriptor`) — review/merge that first.

#32328 fixes the taproot `h`-descriptor bug (#16748) with a spot conversion in `getAccountInfo`. This PR turns that "remember to convert" rule into a **compiler-enforced** one, so no future backend call site can silently reintroduce the bug.

## What it does

- Adds a `CanonicalDescriptor` brand + `toCanonicalDescriptor` normalizer to `@trezor/utils` (co-located with `convertTaprootXpub`). `toCanonicalDescriptor` is idempotent and total — already-canonical descriptors and every non-taproot descriptor (addresses, plain xpubs) pass through unchanged.
- Requires `CanonicalDescriptor` on the connect `Blockchain` wrapper's `getAccountInfo` / `getAccountUtxo`. The type now **forces** every backend caller to normalize; the taproot fix falls out of the type instead of living as a spot fix.
- Rewrites `getAccountInfo` to mint via the typed converter (subsuming #32328's conversion), and normalizes the other backend call sites the wrapper feeds (`discoverAccounts`, `signTransaction`, `sendTransaction`, `solanaComposeTransaction`, `common/Discovery`) — no-ops at runtime for their already-canonical descriptors.

## Scope & safety

- Confined to `@trezor/utils` + `@trezor/connect`. `blockchain-link` transport types are **untouched** (an earlier attempt to brand them leaked into blockchain-link's own workers/tests, so the brand is enforced at the connect boundary instead).
- **Public SDK API is unaffected**: `TrezorConnect.getAccountInfo`/`discoverAccounts` params already `Omit`/`Pick` the descriptor away from the internal transport type, so the brand never reaches the public surface.
- **Echo contract preserved**: `getAccountInfo` still returns the caller's *original* descriptor; only the backend request is canonicalized.

## Verification

- `@trezor/connect` `type-check` green; `@trezor/utils` + `@trezor/connect` `lint:js` (with `ESLINT_RUN_EXPENSIVE_CHECKS=true`) clean.
- `@trezor/utils` unit tests for `toCanonicalDescriptor` (h→apostrophe, idempotence, non-taproot/empty pass-through).

## Possible follow-ups (out of scope here)

- Unify `@suite-common/wallet-types`' existing `AccountDescriptor` brand with `CanonicalDescriptor` and make `asAccountDescriptor` normalize.
- Add `toFirmwareDescriptor` + route the display sites (`ConfirmXpubModal`, native `AccountSettingsShowXpubButton`) through it.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## 🔍 Currents Test Results

🔍 **Suite native android test results:** [View in Currents](https://app.currents.dev/projects/iUe1Y4/runs?branch=feat_canonical_descriptor_brand&lookback=7)

🔍 **Suite desktop test results:** [View in Currents](https://app.currents.dev/projects/4ytF0E/runs?branch=feat_canonical_descriptor_brand&lookback=7)

🔍 **Suite web test results:** [View in Currents](https://app.currents.dev/projects/Og0NOQ/runs?branch=feat_canonical_descriptor_brand&lookback=7)

<details>
<summary><h3>🤖 LLM Test Recommendations</h3></summary>

**Summary:** This change set touches core connect-core transaction APIs (signTransaction, sendTransaction, solanaComposeTransaction), account discovery (Discovery, discoverAccounts, getAccountInfo), the Blockchain backend, and a Bitcoin taproot xpub utility. The highest regression risk is in cross-coin transaction signing/sending and account discovery. The recommended strategy is a focused set covering BTC/ETH/SOL send flows, trading sign flows, discovery resilience, direct TrezorConnect API tests, and taproot account creation.

<details><summary><b>Changed files (9)</b></summary>

- `packages/connect-core/src/api/common/Discovery.ts`
- `packages/connect-core/src/api/discoverAccounts.ts`
- `packages/connect-core/src/api/getAccountInfo.ts`
- `packages/connect-core/src/api/sendTransaction.ts`
- `packages/connect-core/src/api/signTransaction.ts`
- `packages/connect-core/src/api/solana/api/solanaComposeTransaction.ts`
- `packages/connect-core/src/backend/Blockchain.ts`
- `packages/utils/src/convertTaprootXpub.test.ts`
- `packages/utils/src/convertTaprootXpub.ts`

</details>

**Recommended tests (15)**

<details><summary>🔴 <b>High priority (9)</b></summary>

- `suite/e2e/tests/general/wallet-discovery.test.ts` — Directly exercises standard wallet discovery after ejecting/adding a wallet, which is the primary user-facing path affected by changes to Discovery.ts, discoverAccounts.ts, getAccountInfo.ts and Blockchain.ts.
- `suite/e2e/tests/trading/sell-bitcoin.test.ts` — Statically mapped to signTransaction.ts and exercises the full BTC transaction signing path in a trading context, including device prompts and broadcast.
- `suite/e2e/tests/trading/sell-solana.test.ts` — Statically mapped to solanaComposeTransaction.ts and covers Solana sell transaction composition, confirmation, and sending end-to-end.
- `suite/e2e/tests/trezor-connect/getAccountInfo.test.ts` — Directly tests the getAccountInfo API that was changed, including account type selection and successful account info retrieval.
- `suite/e2e/tests/trezor-connect/signTransaction.test.ts` — Directly tests the signTransaction API with BTC inputs, exercising the changed signTransaction.ts implementation and on-device confirmation flow.
- `suite/e2e/tests/wallet/discovery.test.ts` — Stress-tests account discovery across multiple coins with a page reload interruption, directly validating Discovery.ts, discoverAccounts.ts and Blockchain.ts resilience.
- `suite/e2e/tests/wallet/send-eth.test.ts` — Inferred coverage for sendTransaction.ts: exercises the full EVM send flow with custom fees and device confirmation, a path that depends on the changed send transaction implementation.
- `suite/e2e/tests/wallet/send-form-regtest.test.ts` — Statically mapped to signTransaction.ts and provides the most thorough BTC send coverage (multiple outputs, locktime, OP_RETURN), also relying on sendTransaction and Blockchain.
- `suite/e2e/tests/wallet/send-sol.test.ts` — Statically mapped to solanaComposeTransaction.ts and covers Solana send-max transaction composition, fee calculation, device confirmation and send, hitting both solanaComposeTransaction and sendTransaction.

</details>

<details><summary>🟡 <b>Medium priority (6)</b></summary>

- `suite/e2e/tests/trading/swap-fees-bitcoin.test.ts` — Statically mapped to signTransaction.ts; exercises BTC swap signing with custom fees, adding coverage for fee handling in the sign path beyond the sell flow.
- `suite/e2e/tests/trading/swap-sol-to-btc.test.ts` — Statically mapped to solanaComposeTransaction.ts; tests cross-chain SOL-to-BTC swap composition and status progression, complementing the SOL sell test.
- `suite/e2e/tests/trezor-connect/ethereumSignTransaction.test.ts` — Covers the ETH transaction signing flow via TrezorConnect, providing extra confidence for signTransaction.ts changes on an EVM coin.
- `suite/e2e/tests/wallet/add-account-types.test.ts` — Explicitly creates taproot BTC accounts, the specific scenario most likely to exercise the convertTaprootXpub utility change.
- `suite/e2e/tests/wallet/send-doge.test.ts` — Statically mapped to signTransaction.ts; tests a large-amount edge case on a UTXO coin, adding confidence that signTransaction handles unusual DOGE amounts correctly.
- `suite/e2e/tests/wallet/send-form-ltc.test.ts` — Statically mapped to signTransaction.ts; covers LTC send with a MimbleWimble peg-out input, adding UTXO coin diversity for the signing change.

</details>

_Updated: 2026-09-16T12:15:24.472Z_

</details>

## 🔒 Quarantined E2E Tests

<details><summary>Trezor Suite (web) — 5 test(s)</summary>

| Test | Type |
|------|------|
| Quarantine test: "Trading - Swap,Swap SOL USDT token to ETH" | 🙋 manual |
| Quarantine test: "Trading - Swap,Swap SOL USDT token to ETH" | 🙋 manual |
| Quarantine test: "Trading - Buy BTC,Buy Bitcoin from best offer" | 🙋 manual |
| Trading - Swap > Swap SOL USDT token to ETH | 🤖 auto |
| Trading - Swap coin to token > Swap Solana to USDC | 🤖 auto |

</details>

_Updated: 2026-09-16T12:17:32.691Z • 5 test(s) total_

<details><summary>Trezor Suite (desktop) — 1 test(s)</summary>

| Test | Type |
|------|------|
| Trading - Swap > Swap SOL USDT token to ETH | 🤖 auto |

</details>

_Updated: 2026-09-16T12:17:05.570Z • 1 test(s) total_

## 🌐 Preview deployments

🌐 **Suite Web preview:** [https://dev.suite.sldev.cz/suite-web/feat_canonical_descriptor_brand/web/](https://dev.suite.sldev.cz/suite-web/feat_canonical_descriptor_brand/web/)

### CODE REVIEW

File: packages/utils/src/convertTaprootXpub.test.ts [hunk 1]
```diff
@@ -43,3 +43,25 @@ describe('convertTaprootXpub', () => {
...
+
+    it('normalizes the firmware `h` form to the canonical apostrophe form', () => {
+        expect(toCanonicalDescriptor(withH)).toEqual(withApostrophe);
+    });
+
+    it('is idempotent on an already-canonical descriptor', () => {
+        expect(toCanonicalDescriptor(withApostrophe)).toEqual(withApostrophe);
+    });
+
+    it('passes non-taproot descriptors through unchanged (address, plain xpub, empty)', () => {
```
Comment (mroz22) [inline packages/utils/src/convertTaprootXpub.test.ts:61]: `non-taproot descriptors` is correct but not the most exact term. but what is better?

File: packages/connect-core/src/api/getAccountInfo.ts [hunk 2]
```diff
@@ -240,9 +232,14 @@ export default class GetAccountInfo extends AbstractMethod<'getAccountInfo', Req
...
 
+                // Blockbook rejects taproot descriptors that use `h` for hardened parts, so the
+                // backend receives the canonical form. The response keeps the caller's original
+                // `descriptor` below, so the API returns the same format it was given.
+                const canonicalDescriptor = toCanonicalDescriptor(descriptor);
+
                 // get account info from backend
                 const info = await blockchain.getAccountInfo({
-                    descriptor: backendDescriptor,
+                    descriptor: canonicalDescriptor,
```
Comment (mroz22) [inline packages/connect-core/src/api/getAccountInfo.ts]: iz that ok with electrum?

File: packages/utils/src/convertTaprootXpub.test.ts (hunk 1 above)
Comment (mroz22) [reply packages/utils/src/convertTaprootXpub.test.ts]: 🤖 *beep* — terminology audit complete.

You're right that `non-taproot` is imprecise, and it's imprecise in a way that actually misdescribes the function. Taproot is incidental here: `convertTaprootXpub` only rewrites what sits between `[` and `]`, so it would happily normalize `wpkh([fp/84'/0'/0']zpub/<0;1>/*)` too. The real predicate is **"has a BIP-380 key-origin prefix"**, not **"is taproot"**.

So the accurate phrasing for what passes through unchanged is *descriptors with no key-origin prefix*:

```ts
it('passes descriptors without a key-origin prefix through unchanged (address, bare xpub, empty)', () => {
```

and in the doc comment, `every non-taproot descriptor (addresses, plain xpubs)` → `every descriptor without a bracketed key-origin prefix (addresses, bare xpubs)`.

"Bare xpub" also beats "plain xpub" — *bare* is the term BIP-380 itself uses for a key expression with no origin info.

Happy to push that wording if you want it in this PR.

File: packages/connect-core/src/api/getAccountInfo.ts (hunk 2 above)
Comment (mroz22) [reply packages/connect-core/src/api/getAccountInfo.ts]: 🤖 *whirr* — I went and read the electrum worker instead of guessing. Short answer: **yes, and electrum is actually stricter than blockbook here — it needs the canonical form.**

The electrum path for an extended descriptor is `workers/electrum/methods/getAccountInfo.ts` → `addressCache(descriptor, …)` → `utxo-lib`'s `createAddressCache` → `deriveAddresses` → `getXpubOrDescriptorInfo`. That lands in `getDescriptorInfo` (`packages/utxo-lib/src/derivation.ts`), whose parser is:

```ts
descriptor.match(
    /^([a-z]+\()+\[([a-z0-9]{8}(?:\/[0-9]+'?){3,})\]([xyztuv]pub[a-zA-Z0-9]*)\/<0;1>\/\*\)+(#[a-z0-9]{8})?$/,
) || throwError(`Descriptor cannot be parsed: ${descriptor}`);
```

Note `[0-9]+'?` — apostrophe only, no `h`. Ran it against both forms:

| descriptor | parses |
|---|---|
| `tr([71d98c03/86h/0h/0h]xpub…/<0;1>/*)` | ❌ throws `Descriptor cannot be parsed` |
| `tr([71d98c03/86'/0'/0']xpub…/<0;1>/*)` | ✅ |
| `tr([71d98c03/86'/0'/0']xpub…/<0;1>/*)#gvfjd7ak` | ✅ |

So `toCanonicalDescriptor` doesn't just fail to hurt electrum, it's the only thing that makes an `h`-form descriptor work there at all. Same direction as blockbook, one normalizer covers both.

Three adjacent things I checked while I was in there:

1. **Plain addresses are untouched.** Electrum's other branch is `tryGetScripthash(descriptor)` for single addresses. `toCanonicalDescriptor` is identity for anything without exactly one `[`…`]` pair (`convertTaprootXpub` returns `null` → fallback to the original string), so addresses, bare xpubs and solana/eth descriptors reach the backend byte-identical. Same reason ripple/solana/blockbook workers are unaffected.

2. **The checksum stays consistent** on the paths this PR touches. `resolveDescriptorForTaproot` already splits `#<checksum>` into `descriptorChecksum` *and* already does `h-to-apostrophe` before anything reaches `account.descriptor`/`xpubSegwit`. So the conversions added in `Discovery`, `discoverAccounts`, `signTransaction` and `sendTransaction` really are runtime no-ops, exactly as advertised — no descriptor there carries a checksum that could go stale.

3. **The one theoretical wart**: if a *third-party* caller hands `getAccountInfo` an `h`-form descriptor *with* a checksum appended, we rewrite the path and leave `#…` as-is, so the checksum no longer matches the string (BIP-380's `INPUT_CHARSET` scores `'` and `h` differently). Electrum doesn't care — its regex treats the checksum group as optional and never verifies it. Blockbook doesn't appear to verify it either. Flagging it for the record rather than as a blocker; if we ever want it airtight, `toCanonicalDescriptor` could re-run `addDescriptorChecksum` when it rewrote a checksummed descriptor. Out of scope for this PR IMO.

---

Also: branch force-pushed onto current `develop` (`f40e3313f2`). Clean rebase, no conflicts — `#32328` is merged so the stack is down to this single commit. `@trezor/connect-core` + `@trezor/utils` `type-check` green, `convertTaprootXpub` suite 8/8. All `blockchain.getAccountInfo`/`getAccountUtxo` call sites in `connect-core` are still covered by the brand; `develop` didn't add new ones.
